File: black-shop-apollo/apollo-portal/src/main/resources/static/scripts/services/ClusterService.js
appService.service('ClusterService', ['$resource', '$q', function ($resource, $q) {
    var cluster_resource = $resource('', {}, {
        create_cluster: {
            method: 'POST',
            url: 'apps/:appId/envs/:env/clusters'
        },
        load_cluster: {
            method: 'GET',
            url: 'apps/:appId/envs/:env/clusters/:clusterName'
        },
        delete_cluster: {
            method: 'DELETE',
            url: 'apps/:appId/envs/:env/clusters/:clusterName'
        }
    });
    return {
        create_cluster: function (appId, env, cluster) {
            var d = $q.defer();
            cluster_resource.create_cluster({
                                                appId: appId,
                                                env: env
                                            }, cluster,
                                            function (result) {
                                                d.resolve(result);
                                            }, function (result) {
                    d.reject(result);
                });
            return d.promise;
        },
        load_cluster: function (appId, env, clusterName) {
            var d = $q.defer();
            cluster_resource.load_cluster({
                appId: appId,
                env: env,
                clusterName: clusterName
            },
            function (result) {
                d.resolve(result);
            }, function (result) {
                d.reject(result);
            });
            return d.promise;
        },
        delete_cluster: function (appId, env, clusterName) {
            var d = $q.defer();
            cluster_resource.delete_cluster({
                appId: appId,
                env: env,
                clusterName: clusterName
            },
            function (result) {
                d.resolve(result);
            }, function (result) {
                d.reject(result);
            });
            return d.promise;
        }
    }
}]);

File: black-shop-apollo/apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/ClusterService.java
package com.ctrip.framework.apollo.portal.service;

import com.ctrip.framework.apollo.common.dto.ClusterDTO;
import com.ctrip.framework.apollo.common.exception.BadRequestException;
import com.ctrip.framework.apollo.core.enums.Env;
import com.ctrip.framework.apollo.portal.api.AdminServiceAPI;
import com.ctrip.framework.apollo.portal.constant.TracerEventType;
import com.ctrip.framework.apollo.portal.spi.UserInfoHolder;
import com.ctrip.framework.apollo.tracer.Tracer;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class ClusterService {

  @Autowired
  private UserInfoHolder userInfoHolder;
  @Autowired
  private AdminServiceAPI.ClusterAPI clusterAPI;

  public List<ClusterDTO> findClusters(Env env, String appId) {
    return clusterAPI.findClustersByApp(appId, env);
  }

  public ClusterDTO createCluster(Env env, ClusterDTO cluster) {
    if (!clusterAPI.isClusterUnique(cluster.getAppId(), env, cluster.getName())) {
      throw new BadRequestException(String.format("cluster %s already exists.", cluster.getName()));
    }
    ClusterDTO clusterDTO = clusterAPI.create(env, cluster);

    Tracer.logEvent(TracerEventType.CREATE_CLUSTER, cluster.getAppId(), "0", cluster.getName());

    return clusterDTO;
  }

  public void deleteCluster(Env env, String appId, String clusterName){
    clusterAPI.delete(env, appId, clusterName, userInfoHolder.getUser().getUserId());
  }

  public ClusterDTO loadCluster(String appId, Env env, String clusterName){
    return clusterAPI.loadCluster(appId, env, clusterName);
  }

}
